library/core/src/intrinsics.rs

```diff
@@ -3142,3 +3153,43 @@ pub(crate) const fn miri_promise_symbolic_alignment(ptr: *const (), align: usize
 
     const_eval_select((ptr, align), compiletime, runtime);
 }
+
+#[cfg(kani)]
+#[unstable(feature="kani", issue="none")]
+mod verify {
+    use core::{cmp, fmt};
+    use super::*;
+    use crate::kani;
+
+    #[kani::proof_for_contract(typed_swap)]
+    pub fn check_typed_swap_u8() {
+        check_swap::<u8>()
+    }
+
+    #[kani::proof_for_contract(typed_swap)]
+    pub fn check_typed_swap_char() {
+        check_swap::<char>()
+    }
+
+    #[kani::proof_for_contract(typed_swap)]
+    pub fn check_typed_swap_non_zero() {
+        check_swap::<core::num::NonZeroI32>()
+    }
+
+    pub fn check_swap<T: kani::Arbitrary + Copy + cmp::PartialEq + fmt::Debug>() {
+        let mut x = kani::any::<T>();
+        let old_x = x;
+        let x_addr = (&x as *const T).addr();
+
+        let mut y = kani::any::<T>();
+        let old_y = y;
+        let y_addr = (&y as *const T).addr();
+
+        unsafe { typed_swap(&mut x, &mut y) };
+
+        assert_eq!(y, old_x);
+        assert_eq!(x, old_y);
+        assert_eq!(y_addr, (&y as *const T).addr());
+        assert_eq!(x_addr, (&x as *const T).addr());
```

Comment:
How could those addresses possibly have changed?

Reply:
Indeed. That is redundant